// @revolist/revogrid/dist/collection/components/revo-grid/revo-grid.js
import { Component, Prop, h, Watch, Element, Listen, Event, Method, State, Host } from '@stencil/core';
import reduce from 'lodash/reduce';
import each from 'lodash/each';
import ColumnDataProvider from '../../services/column.data.provider';
import { DataProvider } from '../../services/data.provider';
import { getVisibleSourceItem } from '../../store/dataSource/data.store';
import DimensionProvider from '../../services/dimension.provider';
import ViewportProvider from '../../services/viewport.provider';
import ThemeService from '../../themeManager/themeService';
import { applyMixins, timeout } from '../../utils/utils';
import AutoSize from '../../plugins/autoSizeColumn';
import { columnTypes } from '../../store/storeTypes';
import FilterPlugin from '../../plugins/filter/filter.plugin';
import SortingPlugin from '../../plugins/sorting/sorting.plugin';
import ExportFilePlugin from '../../plugins/export/export.plugin';
import GroupingRowPlugin from '../../plugins/groupingRow/grouping.row.plugin';
import { RevoViewPort } from './viewport';
import GridRenderService from './viewport.service';
import { ViewPortSections } from './viewport.section';
import RevogrRowHeaders from '../rowHeaders/revogr-row-headers';
import GridScrollingService from './viewport.scrolling.service';
import { UUID } from '../../utils/consts';
export class RevoGridComponent {
  constructor() {
    /**
     * Defines how many rows/columns should be rendered outside visible area.
     */
    this.frameSize = 1;
    /**
     * Indicates default row size.
     * By default 0, means theme package size will be applied
     */
    this.rowSize = 0;
    /** Indicates default column size. */
    this.colSize = 100;
    /** When true, user can range selection. */
    this.range = false;
    /** When true, grid in read only mode. */
    this.readonly = false;
    /** When true, columns are resizable. */
    this.resize = false;
    /** When true cell focus appear. */
    this.canFocus = true;
    /** When true enable clipboard. */
    this.useClipboard = true;
    /**
     * Columns - defines an array of grid columns.
     * Can be column or grouped column.
     */
    this.columns = [];
    /**
     * Source - defines main data source.
     * Can be an Object or 2 dimensional array([][]);
     * Keys/indexes referenced from columns Prop
     */
    this.source = [];
    /** Pinned top Source: {[T in ColumnProp]: any} - defines pinned top rows data source. */
    this.pinnedTopSource = [];
    /** Pinned bottom Source: {[T in ColumnProp]: any} - defines pinned bottom rows data source. */
    this.pinnedBottomSource = [];
    /** Row properies applied */
    this.rowDefinitions = [];
    /** Custom editors register */
    this.editors = {};
    /** Types
     *  Every type represent multiple column properties
     *  Types will be merged but can be replaced with column properties
     */
    this.columnTypes = {};
    /** Theme name */
    this.theme = 'default';
    /**
     * Row class property
     * Define this property in row object and this will be mapped as row class
     */
    this.rowClass = '';
    /**
     * Autosize config
     * Enable columns autoSize, for more details check @autoSizeColumn plugin
     * By default disabled, hence operation is not resource efficient
     * true to enable with default params (double header separator click for autosize)
     * or provide config
     */
    this.autoSizeColumn = false;
    /**
     * Enables filter plugin
     * Can be boolean
     * Can be filter collection
     */
    this.filter = false;
    /**
     * Trimmed rows
     * Functionality which allows to hide rows from main data set
     * @trimmedRows are physical row indexes to hide
     */
    this.trimmedRows = {};
    /**
     * Enables export plugin
     * Can be boolean
     * Can be export options
     */
    this.exporting = false;
    // --------------------------------------------------------------------------
    //
    //  Private Properties
    //
    // --------------------------------------------------------------------------
    // for internal plugin usage
    this.extraElements = [];
    this.uuid = null;
    /**
     * Plugins
     * Define plugins collection
     */
    this.internalPlugins = [];
  }
  // --------------------------------------------------------------------------
  //
  //  Methods
  //
  // --------------------------------------------------------------------------
  /**
   * Refreshes data viewport.
   * Can be specific part as row or pinned row or 'all' by default.
   */
  async refresh(type = 'all') {
    this.dataProvider.refresh(type);
  }
  /**  Scrolls view port to specified row index */
  async scrollToRow(coordinate = 0) {
    const y = this.dimensionProvider.getViewPortPos({
      coordinate,
      dimension: 'row',
    });
    await this.scrollToCoordinate({ y });
  }
  /** Scrolls view port to specified column index */
  async scrollToColumnIndex(coordinate = 0) {
    const x = this.dimensionProvider.getViewPortPos({
      coordinate,
      dimension: 'col',
    });
    await this.scrollToCoordinate({ x });
  }
  /**  Scrolls view port to specified column prop */
  async scrollToColumnProp(prop) {
    const coordinate = this.columnProvider.getColumnIndexByProp(prop, 'col');
    if (coordinate < 0) {
      // already on the screen
      return;
    }
    const x = this.dimensionProvider.getViewPortPos({
      coordinate,
      dimension: 'col',
    });
    await this.scrollToCoordinate({ x });
  }
  /** Update columns */
  async updateColumns(cols) {
    this.columnProvider.updateColumns(cols);
  }
  /** Add trimmed by type */
  async addTrimmed(trimmed, trimmedType = 'external', type = 'row') {
    const event = this.beforeTrimmed.emit({
      trimmed,
      trimmedType,
      type,
    });
    if (event.defaultPrevented) {
      return event;
    }
    this.dataProvider.setTrimmed({ [trimmedType]: event.detail.trimmed }, type);
    this.afterTrimmed.emit();
    return event;
  }
  /**  Scrolls view port to coordinate */
  async scrollToCoordinate(cell) {
    this.scrollToCell(cell);
  }
  /**  Bring cell to edit mode */
  async setCellEdit(row, prop, rowSource = 'row') {
    const col = ColumnDataProvider.getColumnByProp(this.columns, prop);
    if (!col) {
      return;
    }
    await timeout();
    this.setEdit(row, this.columnProvider.getColumnIndexByProp(prop, 'col'), col.pin || 'col', rowSource);
  }
  /**
   * Register new virtual node inside of grid
   * Used for additional items creation such as plugin elements
   */
  async registerVNode(elements) {
    this.extraElements.push(...elements);
    this.extraElements = [...this.extraElements];
  }
  /**  Get data from source */
  async getSource(type = 'row') {
    return this.dataProvider.stores[type].store.get('source');
  }
  /**
   * Get data from visible part of source
   * Trimmed/filtered rows will be excluded
   * @param type - type of source
   */
  async getVisibleSource(type = 'row') {
    return getVisibleSourceItem(this.dataProvider.stores[type].store);
  }
  /**
   * Provides access to rows internal store observer
   * Can be used for plugin support
   * @param type - type of source
   */
  async getSourceStore(type = 'row') {
    return this.dataProvider.stores[type].store;
  }
  /**
   * Provides access to column internal store observer
   * Can be used for plugin support
   * @param type - type of column
   */
  async getColumnStore(type = 'col') {
    return this.columnProvider.stores[type].store;
  }
  /**
   * Update column sorting
   * @param column - full column details to update
   * @param index - virtual column index
   * @param order - order to apply
   */
  async updateColumnSorting(column, index, order) {
    return this.columnProvider.updateColumnSorting(column, index, order);
  }
  /**
   * Receive all columns in data source
   */
  async getColumns() {
    return this.columnProvider.getColumns();
  }
  /**
   * Clear current grid focus
   */
  async clearFocus() {
    return this.clearFocused();
  }
  /**
   * Get all active plugins instances
   */
  async getPlugins() {
    return [...this.internalPlugins];
  }
  // --------------------------------------------------------------------------
  //
  //  Listeners
  //
  // --------------------------------------------------------------------------
  /** Clear data which is outside of grid container */
  handleOutsideClick({ target }) {
    if (!(target === null || target === void 0 ? void 0 : target.closest(`[${UUID}="${this.uuid}"]`))) {
      this.selectionStoreConnector.clearAll();
    }
  }
  /** DRAG AND DROP */
  onRowDragStarted(e) {
    var _a;
    e.cancelBubble = true;
    const dragStart = this.rowDragStart.emit(e.detail);
    if (dragStart.defaultPrevented) {
      e.preventDefault();
      return;
    }
    (_a = this.orderService) === null || _a === void 0 ? void 0 : _a.start(this.element, Object.assign(Object.assign({}, e.detail), dragStart.detail));
  }
  onRowDragEnd() {
    var _a;
    (_a = this.orderService) === null || _a === void 0 ? void 0 : _a.end();
  }
  onRowDrag({ detail }) {
    var _a;
    (_a = this.orderService) === null || _a === void 0 ? void 0 : _a.move(detail);
  }
  onRowMouseMove(e) {
    var _a;
    e.cancelBubble = true;
    (_a = this.orderService) === null || _a === void 0 ? void 0 : _a.moveTip(e.detail);
  }
  async onBeforeEdit(e) {
    e.cancelBubble = true;
    const { defaultPrevented, detail } = this.beforeEdit.emit(e.detail);
    await timeout();
    // apply data
    if (!defaultPrevented) {
      this.dataProvider.setCellData(detail);
      this.afterEdit.emit(detail);
    }
  }
  onBeforeRangeEdit(e) {
    e.cancelBubble = true;
    const { defaultPrevented } = this.beforeRangeEdit.emit(e.detail);
    if (defaultPrevented) {
      e.preventDefault();
      return;
    }
    this.afterEdit.emit(e.detail);
  }
  onRangeChanged(e) {
    e.cancelBubble = true;
    const beforeRange = this.beforeRange.emit(e.detail);
    if (beforeRange.defaultPrevented) {
      e.preventDefault();
    }
    const beforeFill = this.beforeAutofill.emit(e.detail);
    if (beforeFill.defaultPrevented) {
      return;
    }
  }
  onRowDropped(e) {
    e.cancelBubble = true;
    const { defaultPrevented } = this.rowOrderChanged.emit(e.detail);
    if (defaultPrevented) {
      e.preventDefault();
    }
  }
  onHeaderClick(e) {
    const { defaultPrevented } = this.headerClick.emit(Object.assign(Object.assign({}, e.detail.column), { originalEvent: e.detail.originalEvent }));
    if (defaultPrevented) {
      e.preventDefault();
    }
  }
  onCellFocus(e) {
    e.cancelBubble = true;
    const { defaultPrevented } = this.beforeCellFocus.emit(e.detail);
    if (!this.canFocus || defaultPrevented) {
      e.preventDefault();
    }
  }
  columnChanged(newVal) {
    const columnGather = ColumnDataProvider.getColumns(newVal, 0, this.columnTypes);
    this.beforeColumnsSet.emit(columnGather);
    for (let type of columnTypes) {
      const items = columnGather.columns[type];
      this.dimensionProvider.setRealSize(items.length, type);
      this.dimensionProvider.setColumns(type, ColumnDataProvider.getSizes(items), type !== 'col');
    }
    const columns = this.columnProvider.setColumns(columnGather);
    this.afterColumnsSet.emit({
      columns,
      order: this.columnProvider.order,
    });
  }
  themeChanged(t) {
    this.themeService.register(t);
    this.dimensionProvider.setSettings({ originItemSize: this.themeService.rowSize, frameOffset: this.frameSize || 0 }, 'row');
    this.dimensionProvider.setSettings({ originItemSize: this.colSize, frameOffset: this.frameSize || 0 }, 'col');
  }
  dataChanged(source) {
    let newSource = [...source];
    const beforeSourceSet = this.beforeSourceSet.emit({
      type: 'row',
      source: newSource,
    });
    newSource = beforeSourceSet.detail.source;
    newSource = this.dataProvider.setData(newSource, 'row');
    this.afterSourceSet.emit({
      type: 'row',
      source: newSource,
    });
  }
  dataBottomChanged(newVal) {
    this.dataProvider.setData(newVal, 'rowPinEnd');
  }
  dataTopChanged(newVal) {
    this.dataProvider.setData(newVal, 'rowPinStart');
  }
  rowDefChanged(newVal) {
    if (!newVal.length) {
      return;
    }
    const rows = reduce(newVal, (r, v) => {
      if (!r[v.type]) {
        r[v.type] = {};
      }
      if (v.size) {
        if (!r[v.type].sizes) {
          r[v.type].sizes = {};
        }
        r[v.type].sizes[v.index] = v.size;
      }
      return r;
    }, {});
    each(rows, (r, k) => {
      if (r.sizes) {
        this.dimensionProvider.setDimensionSize(k, r.sizes);
      }
    });
  }
  trimmedRowsChanged(newVal) {
    this.addTrimmed(newVal);
  }
  groupingChanged(newVal = {}) {
    let grPlugin;
    for (let p of this.internalPlugins) {
      const isGrouping = p;
      if (isGrouping.setGrouping) {
        grPlugin = isGrouping;
        break;
      }
    }
    if (!grPlugin) {
      return;
    }
    grPlugin.setGrouping(newVal || {});
  }
  connectedCallback() {
    this.viewportProvider = new ViewportProvider();
    this.themeService = new ThemeService({
      rowSize: this.rowSize,
    });
    this.dimensionProvider = new DimensionProvider(this.viewportProvider);
    this.columnProvider = new ColumnDataProvider();
    this.dataProvider = new DataProvider(this.dimensionProvider);
    this.uuid = `${new Date().getTime()}-rvgrid`;
    if (this.autoSizeColumn) {
      this.internalPlugins.push(new AutoSize(this.element, {
        dataProvider: this.dataProvider,
        columnProvider: this.columnProvider,
        dimensionProvider: this.dimensionProvider,
      }, typeof this.autoSizeColumn === 'object' ? this.autoSizeColumn : undefined));
    }
    if (this.filter) {
      this.internalPlugins.push(new FilterPlugin(this.element, this.uuid, typeof this.filter === 'object' ? this.filter : undefined));
    }
    if (this.exporting) {
      this.internalPlugins.push(new ExportFilePlugin(this.element));
    }
    this.internalPlugins.push(new SortingPlugin(this.element));
    if (this.plugins) {
      this.plugins.forEach(p => {
        this.internalPlugins.push(new p(this.element));
      });
    }
    this.internalPlugins.push(new GroupingRowPlugin(this.element, {
      dataProvider: this.dataProvider,
      columnProvider: this.columnProvider,
    }));
    this.themeChanged(this.theme);
    this.columnChanged(this.columns);
    this.dataChanged(this.source);
    this.dataTopChanged(this.pinnedTopSource);
    this.dataBottomChanged(this.pinnedBottomSource);
    this.trimmedRowsChanged(this.trimmedRows);
    this.rowDefChanged(this.rowDefinitions);
    this.groupingChanged(this.grouping);
    this.viewportConnectedCallback();
    this.scrollingService = new GridScrollingService((e) => {
      this.dimensionProvider.setViewPortCoordinate({
        coordinate: e.coordinate,
        type: e.dimension,
      });
      this.viewportScroll.emit(e);
    });
  }
  disconnectedCallback() {
    // destroy plugins on element disconnect
    each(this.internalPlugins, p => p.destroy());
    this.internalPlugins = [];
  }
  render() {
    var _a, _b;
    (_a = this.selectionStoreConnector) === null || _a === void 0 ? void 0 : _a.beforeUpdate();
    const contentHeight = this.dimensionProvider.stores['row'].store.get('realSize');
    const columns = this.getViewportColumnData(contentHeight);
    (_b = this.scrollingService) === null || _b === void 0 ? void 0 : _b.unregister();
    return (h(Host, Object.assign({}, { [`${UUID}`]: this.uuid }),
      h(RevoViewPort, { viewports: this.viewportProvider.stores, dimensions: this.dimensionProvider.stores, orderRef: e => (this.orderService = e), registerElement: (e, k) => this.scrollingService.registerElement(e, k), onScroll: details => this.scrollingService.onScroll(details) },
        this.rowHeaders ? (h(RevogrRowHeaders, { height: contentHeight, anyView: columns[0], resize: this.resize, rowHeaderColumn: typeof this.rowHeaders === 'object' ? this.rowHeaders : undefined, beforeRowAdd: y => this.selectionStoreConnector.registerRow(y), onScrollViewport: e => this.scrollingService.onScroll(e, 'headerRow'), onElementToScroll: e => this.scrollingService.registerElement(e, 'headerRow') })) : (''),
        h(ViewPortSections, { columnFilter: !!this.filter, resize: this.resize, readonly: this.readonly, range: this.range, rowClass: this.rowClass, editors: this.editors, useClipboard: this.useClipboard, columns: columns, onEdit: detail => {
            const event = this.beforeEditStart.emit(detail);
            if (!event.defaultPrevented) {
              this.selectionStoreConnector.setEdit(detail.isCancel ? false : detail.val);
            }
          }, registerElement: (e, k) => this.scrollingService.registerElement(e, k), onScroll: details => this.scrollingService.onScroll(details) })),
      this.extraElements));
  }
  static get is() { return "revo-grid"; }
  static get originalStyleUrls() { return {
    "$": ["revo-grid.scss"]
  }; }
  static get styleUrls() { return {
    "$": ["revo-grid.css"]
  }; }
  static get properties() { return {
    "rowHeaders": {
      "type": "boolean",
      "mutable": false,
      "complexType": {
        "original": "RevoGrid.RowHeaders | boolean",
        "resolved": "RowHeaders | boolean",
        "references": {
          "RevoGrid": {
            "location": "import",
            "path": "../../interfaces"
          }
        }
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "Excel like show row indexe per row"
      },
      "attribute": "row-headers",
      "reflect": false
    },
    "frameSize": {
      "type": "number",
      "mutable": false,
      "complexType": {
        "original": "number",
        "resolved": "number",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "Defines how many rows/columns should be rendered outside visible area."
      },
      "attribute": "frame-size",
      "reflect": false,
      "defaultValue": "1"
    },
    "rowSize": {
      "type": "number",
      "mutable": false,
      "complexType": {
        "original": "number",
        "resolved": "number",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "Indicates default row size.\nBy default 0, means theme package size will be applied"
      },
      "attribute": "row-size",
      "reflect": false,
      "defaultValue": "0"
    },
    "colSize": {
      "type": "number",
      "mutable": false,
      "complexType": {
        "original": "number",
        "resolved": "number",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "Indicates default column size."
      },
      "attribute": "col-size",
      "reflect": false,
      "defaultValue": "100"
    },
    "range": {
      "type": "boolean",
      "mutable": false,
      "complexType": {
        "original": "boolean",
        "resolved": "boolean",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "When true, user can range selection."
      },
      "attribute": "range",
      "reflect": false,
      "defaultValue": "false"
    },
    "readonly": {
      "type": "boolean",
      "mutable": false,
      "complexType": {
        "original": "boolean",
        "resolved": "boolean",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "When true, grid in read only mode."
      },
      "attribute": "readonly",
      "reflect": false,
      "defaultValue": "false"
    },
    "resize": {
      "type": "boolean",
      "mutable": false,
      "complexType": {
        "original": "boolean",
        "resolved": "boolean",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "When true, columns are resizable."
      },
      "attribute": "resize",
      "reflect": false,
      "defaultValue": "false"
    },
    "canFocus": {
      "type": "boolean",
      "mutable": false,
      "complexType": {
        "original": "boolean",
        "resolved": "boolean",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "When true cell focus appear."
      },
      "attribute": "can-focus",
      "reflect": false,
      "defaultValue": "true"
    },
    "useClipboard": {
      "type": "boolean",
      "mutable": false,
      "complexType": {
        "original": "boolean",
        "resolved": "boolean",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "When true enable clipboard."
      },
      "attribute": "use-clipboard",
      "reflect": false,
      "defaultValue": "true"
    },
    "columns": {
      "type": "unknown",
      "mutable": false,
      "complexType": {
        "original": "(RevoGrid.ColumnRegular | RevoGrid.ColumnGrouping)[]",
        "resolved": "(ColumnRegular | ColumnGrouping)[]",
        "references": {
          "RevoGrid": {
            "location": "import",
            "path": "../../interfaces"
          }
        }
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "Columns - defines an array of grid columns.\nCan be column or grouped column."
      },
      "defaultValue": "[]"
    },
    "source": {
      "type": "unknown",
      "mutable": false,
      "complexType": {
        "original": "RevoGrid.DataType[]",
        "resolved": "DataType[]",
        "references": {
          "RevoGrid": {
            "location": "import",
            "path": "../../interfaces"
          }
        }
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "Source - defines main data source.\nCan be an Object or 2 dimensional array([][]);\nKeys/indexes referenced from columns Prop"
      },
      "defaultValue": "[]"
    },
    "pinnedTopSource": {
      "type": "unknown",
      "mutable": false,
      "complexType": {
        "original": "RevoGrid.DataType[]",
        "resolved": "DataType[]",
        "references": {
          "RevoGrid": {
            "location": "import",
            "path": "../../interfaces"
          }
        }
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "Pinned top Source: {[T in ColumnProp]: any} - defines pinned top rows data source."
      },
      "defaultValue": "[]"
    },
    "pinnedBottomSource": {
      "type": "unknown",
      "mutable": false,
      "complexType": {
        "original": "RevoGrid.DataType[]",
        "resolved": "DataType[]",
        "references": {
          "RevoGrid": {
            "location": "import",
            "path": "../../interfaces"
          }
        }
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "Pinned bottom Source: {[T in ColumnProp]: any} - defines pinned bottom rows data source."
      },
      "defaultValue": "[]"
    },
    "rowDefinitions": {
      "type": "unknown",
      "mutable": false,
      "complexType": {
        "original": "RevoGrid.RowDefinition[]",
        "resolved": "RowDefinition[]",
        "references": {
          "RevoGrid": {
            "location": "import",
            "path": "../../interfaces"
          }
        }
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "Row properies applied"
      },
      "defaultValue": "[]"
    },
    "editors": {
      "type": "unknown",
      "mutable": false,
      "complexType": {
        "original": "Edition.Editors",
        "resolved": "{ [name: string]: EditorCtr; }",
        "references": {
          "Edition": {
            "location": "import",
            "path": "../../interfaces"
          }
        }
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "Custom editors register"
      },
      "defaultValue": "{}"
    },
    "plugins": {
      "type": "unknown",
      "mutable": false,
      "complexType": {
        "original": "RevoPlugin.PluginClass[]",
        "resolved": "(typeof Plugin)[]",
        "references": {
          "RevoPlugin": {
            "location": "import",
            "path": "../../interfaces"
          }
        }
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "Custom grid plugins\nHas to be predefined during first grid init\nEvery plugin should be inherited from BasePlugin"
      }
    },
    "columnTypes": {
      "type": "unknown",
      "mutable": false,
      "complexType": {
        "original": "{ [name: string]: RevoGrid.ColumnType }",
        "resolved": "{ [name: string]: ColumnType; }",
        "references": {
          "RevoGrid": {
            "location": "import",
            "path": "../../interfaces"
          }
        }
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "Types\nEvery type represent multiple column properties\nTypes will be merged but can be replaced with column properties"
      },
      "defaultValue": "{}"
    },
    "theme": {
      "type": "string",
      "mutable": true,
      "complexType": {
        "original": "ThemeSpace.Theme",
        "resolved": "\"compact\" | \"darkCompact\" | \"darkMaterial\" | \"default\" | \"material\"",
        "references": {
          "ThemeSpace": {
            "location": "import",
            "path": "../../interfaces"
          }
        }
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "Theme name"
      },
      "attribute": "theme",
      "reflect": true,
      "defaultValue": "'default'"
    },
    "rowClass": {
      "type": "string",
      "mutable": false,
      "complexType": {
        "original": "string",
        "resolved": "string",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "Row class property\nDefine this property in row object and this will be mapped as row class"
      },
      "attribute": "row-class",
      "reflect": true,
      "defaultValue": "''"
    },
    "autoSizeColumn": {
      "type": "boolean",
      "mutable": false,
      "complexType": {
        "original": "boolean | AutoSizeColumnConfig",
        "resolved": "boolean | { mode?: ColumnAutoSizeMode; allColumns?: boolean; letterBlockSize?: number; preciseSize?: boolean; }",
        "references": {
          "AutoSizeColumnConfig": {
            "location": "import",
            "path": "../../plugins/autoSizeColumn"
          }
        }
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "Autosize config\nEnable columns autoSize, for more details check @autoSizeColumn plugin\nBy default disabled, hence operation is not resource efficient\ntrue to enable with default params (double header separator click for autosize)\nor provide config"
      },
      "attribute": "auto-size-column",
      "reflect": false,
      "defaultValue": "false"
    },
    "filter": {
      "type": "boolean",
      "mutable": false,
      "complexType": {
        "original": "boolean | ColumnFilterConfig",
        "resolved": "boolean | { collection?: Record<ColumnProp, FilterCollectionItem>; include?: string[]; customFilters?: Record<string, CustomFilter>; }",
        "references": {
          "ColumnFilterConfig": {
            "location": "import",
            "path": "../../plugins/filter/filter.plugin"
          }
        }
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "Enables filter plugin\nCan be boolean\nCan be filter collection"
      },
      "attribute": "filter",
      "reflect": false,
      "defaultValue": "false"
    },
    "trimmedRows": {
      "type": "unknown",
      "mutable": false,
      "complexType": {
        "original": "Record<number, boolean>",
        "resolved": "{ [x: number]: boolean; }",
        "references": {
          "Record": {
            "location": "global"
          }
        }
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [{
            "text": "are physical row indexes to hide",
            "name": "trimmedRows"
          }],
        "text": "Trimmed rows\nFunctionality which allows to hide rows from main data set"
      },
      "defaultValue": "{}"
    },
    "exporting": {
      "type": "boolean",
      "mutable": false,
      "complexType": {
        "original": "boolean",
        "resolved": "boolean",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "Enables export plugin\nCan be boolean\nCan be export options"
      },
      "attribute": "exporting",
      "reflect": false,
      "defaultValue": "false"
    },
    "grouping": {
      "type": "unknown",
      "mutable": false,
      "complexType": {
        "original": "GroupingOptions",
        "resolved": "{ props?: ColumnProp[]; expandedAll?: boolean; }",
        "references": {
          "GroupingOptions": {
            "location": "import",
            "path": "../../plugins/groupingRow/grouping.row.types"
          }
        }
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "Group models by provided properties\nDefine properties to be groped by"
      }
    }
  }; }
  static get states() { return {
    "extraElements": {}
  }; }
  static get events() { return [{
      "method": "beforeEdit",
      "name": "beforeEdit",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [],
        "text": "Before edit event.\nTriggered before edit data applied.\nUse e.preventDefault() to prevent edit data set and use you own.\nUse e.val = {your value} to replace edit result with your own."
      },
      "complexType": {
        "original": "Edition.BeforeSaveDataDetails",
        "resolved": "{ prop: ColumnProp; model: DataType; val?: string; rowIndex: number; type: DimensionRows; }",
        "references": {
          "Edition": {
            "location": "import",
            "path": "../../interfaces"
          }
        }
      }
    }, {
      "method": "beforeRangeEdit",
      "name": "beforeRangeEdit",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [],
        "text": "Before range edit event.\nTriggered before range data applied, when range selection happened.\nUse e.preventDefault() to prevent edit data set and use you own."
      },
      "complexType": {
        "original": "Edition.BeforeRangeSaveDataDetails",
        "resolved": "{ data: DataLookup; models: { [rowIndex: number]: DataType; }; type: DimensionRows; }",
        "references": {
          "Edition": {
            "location": "import",
            "path": "../../interfaces"
          }
        }
      }
    }, {
      "method": "afterEdit",
      "name": "afterEdit",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [],
        "text": "After edit.\nTriggered when after data applied or Range changeged."
      },
      "complexType": {
        "original": "Edition.BeforeSaveDataDetails | Edition.BeforeRangeSaveDataDetails",
        "resolved": "{ data: DataLookup; models: { [rowIndex: number]: DataType; }; type: DimensionRows; } | { prop: ColumnProp; model: DataType; val?: string; rowIndex: number; type: DimensionRows; }",
        "references": {
          "Edition": {
            "location": "import",
            "path": "../../interfaces"
          }
        }
      }
    }, {
      "method": "beforeAutofill",
      "name": "beforeAutofill",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [],
        "text": "Before autofill.\nTriggered before autofill applied.\nUse e.preventDefault() to prevent edit data apply."
      },
      "complexType": {
        "original": "Selection.ChangedRange",
        "resolved": "{ type: DimensionRows; newRange: RangeArea; oldRange: RangeArea; newProps: ColumnProp[]; oldProps: ColumnProp[]; newData: { [key: number]: DataType; }; }",
        "references": {
          "Selection": {
            "location": "import",
            "path": "../../interfaces"
          }
        }
      }
    }, {
      "method": "beforeRange",
      "name": "beforeRange",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [],
        "text": "Before range apply.\nTriggered before range applied.\nUse e.preventDefault() to prevent range."
      },
      "complexType": {
        "original": "Selection.ChangedRange",
        "resolved": "{ type: DimensionRows; newRange: RangeArea; oldRange: RangeArea; newProps: ColumnProp[]; oldProps: ColumnProp[]; newData: { [key: number]: DataType; }; }",
        "references": {
          "Selection": {
            "location": "import",
            "path": "../../interfaces"
          }
        }
      }
    }, {
      "method": "rowOrderChanged",
      "name": "rowOrderChanged",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [],
        "text": "Before row order apply.\nUse e.preventDefault() to prevent row order change."
      },
      "complexType": {
        "original": "{ from: number; to: number }",
        "resolved": "{ from: number; to: number; }",
        "references": {}
      }
    }, {
      "method": "beforeSourceSortingApply",
      "name": "beforeSourceSortingApply",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [],
        "text": "Before source update sorting apply.\nUse this event if you intended to prevent sorting on data update.\nUse e.preventDefault() to prevent sorting data change during rows source update."
      },
      "complexType": {
        "original": "any",
        "resolved": "any",
        "references": {}
      }
    }, {
      "method": "beforeSortingApply",
      "name": "beforeSortingApply",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [],
        "text": "Before sorting apply.\nUse e.preventDefault() to prevent sorting data change."
      },
      "complexType": {
        "original": "{\n    column: RevoGrid.ColumnRegular;\n    order: 'desc' | 'asc';\n  }",
        "resolved": "{ column: ColumnRegular; order: \"desc\" | \"asc\"; }",
        "references": {
          "RevoGrid": {
            "location": "import",
            "path": "../../interfaces"
          }
        }
      }
    }, {
      "method": "beforeSorting",
      "name": "beforeSorting",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [],
        "text": "Before sorting event.\nInitial sorting triggered, if this event stops no other event called.\nUse e.preventDefault() to prevent sorting."
      },
      "complexType": {
        "original": "{\n    column: RevoGrid.ColumnRegular;\n    order: 'desc' | 'asc';\n  }",
        "resolved": "{ column: ColumnRegular; order: \"desc\" | \"asc\"; }",
        "references": {
          "RevoGrid": {
            "location": "import",
            "path": "../../interfaces"
          }
        }
      }
    }, {
      "method": "rowDragStart",
      "name": "rowDragStart",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [],
        "text": "Row order change started.\nUse e.preventDefault() to prevent row order change.\nUse e.text = 'new name' to change item name on start."
      },
      "complexType": {
        "original": "{ pos: RevoGrid.PositionItem; text: string }",
        "resolved": "{ pos: PositionItem; text: string; }",
        "references": {
          "RevoGrid": {
            "location": "import",
            "path": "../../interfaces"
          }
        }
      }
    }, {
      "method": "headerClick",
      "name": "headerClick",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [],
        "text": "On header click."
      },
      "complexType": {
        "original": "RevoGrid.ColumnRegular",
        "resolved": "ColumnRegular",
        "references": {
          "RevoGrid": {
            "location": "import",
            "path": "../../interfaces"
          }
        }
      }
    }, {
      "method": "beforeCellFocus",
      "name": "beforeCellFocus",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [],
        "text": "Before cell focus changed.\nUse e.preventDefault() to prevent cell focus change."
      },
      "complexType": {
        "original": "Edition.BeforeSaveDataDetails",
        "resolved": "{ prop: ColumnProp; model: DataType; val?: string; rowIndex: number; type: DimensionRows; }",
        "references": {
          "Edition": {
            "location": "import",
            "path": "../../interfaces"
          }
        }
      }
    }, {
      "method": "beforeSourceSet",
      "name": "beforeSourceSet",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [],
        "text": "Before data apply.\nYou can override data source here"
      },
      "complexType": {
        "original": "{\n    type: RevoGrid.DimensionRows;\n    source: RevoGrid.DataType[];\n  }",
        "resolved": "{ type: DimensionRows; source: DataType[]; }",
        "references": {
          "RevoGrid": {
            "location": "import",
            "path": "../../interfaces"
          }
        }
      }
    }, {
      "method": "afterSourceSet",
      "name": "afterSourceSet",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [],
        "text": "After rows updated"
      },
      "complexType": {
        "original": "{\n    type: RevoGrid.DimensionRows;\n    source: RevoGrid.DataType[];\n  }",
        "resolved": "{ type: DimensionRows; source: DataType[]; }",
        "references": {
          "RevoGrid": {
            "location": "import",
            "path": "../../interfaces"
          }
        }
      }
    }, {
      "method": "beforeColumnsSet",
      "name": "beforeColumnsSet",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [],
        "text": "Before column update"
      },
      "complexType": {
        "original": "ColumnCollection",
        "resolved": "{ columns: Record<DimensionCols, ColumnRegular[]>; columnGrouping: Record<DimensionCols, Group[]>; maxLevel: number; sort: Record<ColumnProp, ColumnRegular>; }",
        "references": {
          "ColumnCollection": {
            "location": "import",
            "path": "../../services/column.data.provider"
          }
        }
      }
    }, {
      "method": "afterColumnsSet",
      "name": "afterColumnsSet",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [],
        "text": "Column updated"
      },
      "complexType": {
        "original": "{\n    columns: ColumnCollection;\n    order: Record<RevoGrid.ColumnProp, 'asc' | 'desc'>;\n  }",
        "resolved": "{ columns: ColumnCollection; order: Record<ColumnProp, \"desc\" | \"asc\">; }",
        "references": {
          "ColumnCollection": {
            "location": "import",
            "path": "../../services/column.data.provider"
          },
          "Record": {
            "location": "global"
          },
          "RevoGrid": {
            "location": "import",
            "path": "../../interfaces"
          }
        }
      }
    }, {
      "method": "beforeFilterApply",
      "name": "beforeFilterApply",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [],
        "text": "Before filter applied to data source\nUse e.preventDefault() to prevent cell focus change\nUpdate @collection if you wish to change filters"
      },
      "complexType": {
        "original": "{ collection: FilterCollection }",
        "resolved": "{ collection: Record<ColumnProp, FilterCollectionItem>; }",
        "references": {
          "FilterCollection": {
            "location": "import",
            "path": "../../plugins/filter/filter.plugin"
          }
        }
      }
    }, {
      "method": "beforeFilterTrimmed",
      "name": "beforeFilterTrimmed",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [],
        "text": "Before filter trimmed values\nUse e.preventDefault() to prevent value trimming and filter apply\nUpdate @collection if you wish to change filters\nUpdate @itemsToFilter if you wish to filter indexes of trimming"
      },
      "complexType": {
        "original": "{ collection: FilterCollection; itemsToFilter: Record<number, boolean> }",
        "resolved": "{ collection: Record<ColumnProp, FilterCollectionItem>; itemsToFilter: Record<number, boolean>; }",
        "references": {
          "FilterCollection": {
            "location": "import",
            "path": "../../plugins/filter/filter.plugin"
          },
          "Record": {
            "location": "global"
          }
        }
      }
    }, {
      "method": "beforeTrimmed",
      "name": "beforeTrimmed",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [],
        "text": "Before trimmed values\nUse e.preventDefault() to prevent value trimming\nUpdate @trimmed if you wish to filter indexes of trimming"
      },
      "complexType": {
        "original": "{ trimmed: Record<number, boolean>; trimmedType: string; type: string }",
        "resolved": "{ trimmed: Record<number, boolean>; trimmedType: string; type: string; }",
        "references": {
          "Record": {
            "location": "global"
          }
        }
      }
    }, {
      "method": "afterTrimmed",
      "name": "afterTrimmed",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [],
        "text": "Notify trimmed applied"
      },
      "complexType": {
        "original": "any",
        "resolved": "any",
        "references": {}
      }
    }, {
      "method": "viewportScroll",
      "name": "viewportScroll",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [],
        "text": "Triggered when view port scrolled"
      },
      "complexType": {
        "original": "RevoGrid.ViewPortScrollEvent",
        "resolved": "{ dimension: DimensionType; coordinate: number; delta?: number; }",
        "references": {
          "RevoGrid": {
            "location": "import",
            "path": "../../interfaces"
          }
        }
      }
    }, {
      "method": "beforeExport",
      "name": "beforeExport",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [],
        "text": "Before export\nUse e.preventDefault() to prevent export\nReplace data in Event in case you want to modify it in export"
      },
      "complexType": {
        "original": "DataInput",
        "resolved": "{ data: DataType[]; } & ColSource",
        "references": {
          "DataInput": {
            "location": "import",
            "path": "../../plugins/export/types"
          }
        }
      }
    }, {
      "method": "beforeEditStart",
      "name": "beforeEditStart",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [],
        "text": "Before edit started\nUse e.preventDefault() to prevent edit"
      },
      "complexType": {
        "original": "Edition.BeforeSaveDataDetails",
        "resolved": "{ prop: ColumnProp; model: DataType; val?: string; rowIndex: number; type: DimensionRows; }",
        "references": {
          "Edition": {
            "location": "import",
            "path": "../../interfaces"
          }
        }
      }
    }]; }
  static get methods() { return {
    "refresh": {
      "complexType": {
        "signature": "(type?: RevoGrid.DimensionRows | 'all') => Promise<void>",
        "parameters": [{
            "tags": [],
            "text": ""
          }],
        "references": {
          "Promise": {
            "location": "global"
          },
          "RevoGrid": {
            "location": "import",
            "path": "../../interfaces"
          }
        },
        "return": "Promise<void>"
      },
      "docs": {
        "text": "Refreshes data viewport.\nCan be specific part as row or pinned row or 'all' by default.",
        "tags": []
      }
    },
    "scrollToRow": {
      "complexType": {
        "signature": "(coordinate?: number) => Promise<void>",
        "parameters": [{
            "tags": [],
            "text": ""
          }],
        "references": {
          "Promise": {
            "location": "global"
          }
        },
        "return": "Promise<void>"
      },
      "docs": {
        "text": "Scrolls view port to specified row index",
        "tags": []
      }
    },
    "scrollToColumnIndex": {
      "complexType": {
        "signature": "(coordinate?: number) => Promise<void>",
        "parameters": [{
            "tags": [],
            "text": ""
          }],
        "references": {
          "Promise": {
            "location": "global"
          }
        },
        "return": "Promise<void>"
      },
      "docs": {
        "text": "Scrolls view port to specified column index",
        "tags": []
      }
    },
    "scrollToColumnProp": {
      "complexType": {
        "signature": "(prop: RevoGrid.ColumnProp) => Promise<void>",
        "parameters": [{
            "tags": [],
            "text": ""
          }],
        "references": {
          "Promise": {
            "location": "global"
          },
          "RevoGrid": {
            "location": "import",
            "path": "../../interfaces"
          }
        },
        "return": "Promise<void>"
      },
      "docs": {
        "text": "Scrolls view port to specified column prop",
        "tags": []
      }
    },
    "updateColumns": {
      "complexType": {
        "signature": "(cols: RevoGrid.ColumnRegular[]) => Promise<void>",
        "parameters": [{
            "tags": [],
            "text": ""
          }],
        "references": {
          "Promise": {
            "location": "global"
          },
          "RevoGrid": {
            "location": "import",
            "path": "../../interfaces"
          }
        },
        "return": "Promise<void>"
      },
      "docs": {
        "text": "Update columns",
        "tags": []
      }
    },
    "addTrimmed": {
      "complexType": {
        "signature": "(trimmed: Record<number, boolean>, trimmedType?: string, type?: RevoGrid.DimensionRows) => Promise<CustomEvent<{ trimmed: Record<number, boolean>; trimmedType: string; type: string; }>>",
        "parameters": [{
            "tags": [],
            "text": ""
          }, {
            "tags": [],
            "text": ""
          }, {
            "tags": [],
            "text": ""
          }],
        "references": {
          "Promise": {
            "location": "global"
          },
          "CustomEvent": {
            "location": "global"
          },
          "Record": {
            "location": "global"
          },
          "RevoGrid": {
            "location": "import",
            "path": "../../interfaces"
          }
        },
        "return": "Promise<CustomEvent<{ trimmed: Record<number, boolean>; trimmedType: string; type: string; }>>"
      },
      "docs": {
        "text": "Add trimmed by type",
        "tags": []
      }
    },
    "scrollToCoordinate": {
      "complexType": {
        "signature": "(cell: Partial<Selection.Cell>) => Promise<void>",
        "parameters": [{
            "tags": [],
            "text": ""
          }],
        "references": {
          "Promise": {
            "location": "global"
          },
          "Partial": {
            "location": "global"
          },
          "Selection": {
            "location": "import",
            "path": "../../interfaces"
          }
        },
        "return": "Promise<void>"
      },
      "docs": {
        "text": "Scrolls view port to coordinate",
        "tags": []
      }
    },
    "setCellEdit": {
      "complexType": {
        "signature": "(row: number, prop: RevoGrid.ColumnProp, rowSource?: RevoGrid.DimensionRows) => Promise<void>",
        "parameters": [{
            "tags": [],
            "text": ""
          }, {
            "tags": [],
            "text": ""
          }, {
            "tags": [],
            "text": ""
          }],
        "references": {
          "Promise": {
            "location": "global"
          },
          "RevoGrid": {
            "location": "import",
            "path": "../../interfaces"
          }
        },
        "return": "Promise<void>"
      },
      "docs": {
        "text": "Bring cell to edit mode",
        "tags": []
      }
    },
    "registerVNode": {
      "complexType": {
        "signature": "(elements: VNode[]) => Promise<void>",
        "parameters": [{
            "tags": [],
            "text": ""
          }],
        "references": {
          "Promise": {
            "location": "global"
          },
          "VNode": {
            "location": "import",
            "path": "@stencil/core"
          }
        },
        "return": "Promise<void>"
      },
      "docs": {
        "text": "Register new virtual node inside of grid\nUsed for additional items creation such as plugin elements",
        "tags": []
      }
    },
    "getSource": {
      "complexType": {
        "signature": "(type?: RevoGrid.DimensionRows) => Promise<RevoGrid.DataType[]>",
        "parameters": [{
            "tags": [],
            "text": ""
          }],
        "references": {
          "Promise": {
            "location": "global"
          },
          "RevoGrid": {
            "location": "import",
            "path": "../../interfaces"
          }
        },
        "return": "Promise<DataType[]>"
      },
      "docs": {
        "text": "Get data from source",
        "tags": []
      }
    },
    "getVisibleSource": {
      "complexType": {
        "signature": "(type?: RevoGrid.DimensionRows) => Promise<any[]>",
        "parameters": [{
            "tags": [{
                "text": "type - type of source",
                "name": "param"
              }],
            "text": "- type of source"
          }],
        "references": {
          "Promise": {
            "location": "global"
          },
          "RevoGrid": {
            "location": "import",
            "path": "../../interfaces"
          }
        },
        "return": "Promise<any[]>"
      },
      "docs": {
        "text": "Get data from visible part of source\nTrimmed/filtered rows will be excluded",
        "tags": [{
            "name": "param",
            "text": "type - type of source"
          }]
      }
    },
    "getSourceStore": {
      "complexType": {
        "signature": "(type?: RevoGrid.DimensionRows) => Promise<RowSource>",
        "parameters": [{
            "tags": [{
                "text": "type - type of source",
                "name": "param"
              }],
            "text": "- type of source"
          }],
        "references": {
          "Promise": {
            "location": "global"
          },
          "RevoGrid": {
            "location": "import",
            "path": "../../interfaces"
          },
          "RowSource": {
            "location": "import",
            "path": "../data/columnService"
          }
        },
        "return": "Promise<Observable<DataSourceState<DataType, DimensionRows>>>"
      },
      "docs": {
        "text": "Provides access to rows internal store observer\nCan be used for plugin support",
        "tags": [{
            "name": "param",
            "text": "type - type of source"
          }]
      }
    },
    "getColumnStore": {
      "complexType": {
        "signature": "(type?: RevoGrid.DimensionCols) => Promise<ColumnSource>",
        "parameters": [{
            "tags": [{
                "text": "type - type of column",
                "name": "param"
              }],
            "text": "- type of column"
          }],
        "references": {
          "Promise": {
            "location": "global"
          },
          "RevoGrid": {
            "location": "import",
            "path": "../../interfaces"
          },
          "ColumnSource": {
            "location": "import",
            "path": "../data/columnService"
          }
        },
        "return": "Promise<Observable<DataSourceState<ColumnRegular, DimensionCols>>>"
      },
      "docs": {
        "text": "Provides access to column internal store observer\nCan be used for plugin support",
        "tags": [{
            "name": "param",
            "text": "type - type of column"
          }]
      }
    },
    "updateColumnSorting": {
      "complexType": {
        "signature": "(column: RevoGrid.ColumnRegular, index: number, order: 'asc' | 'desc') => Promise<RevoGrid.ColumnRegular>",
        "parameters": [{
            "tags": [{
                "text": "column - full column details to update",
                "name": "param"
              }],
            "text": "- full column details to update"
          }, {
            "tags": [{
                "text": "index - virtual column index",
                "name": "param"
              }],
            "text": "- virtual column index"
          }, {
            "tags": [{
                "text": "order - order to apply",
                "name": "param"
              }],
            "text": "- order to apply"
          }],
        "references": {
          "Promise": {
            "location": "global"
          },
          "RevoGrid": {
            "location": "import",
            "path": "../../interfaces"
          }
        },
        "return": "Promise<ColumnRegular>"
      },
      "docs": {
        "text": "Update column sorting",
        "tags": [{
            "name": "param",
            "text": "column - full column details to update"
          }, {
            "name": "param",
            "text": "index - virtual column index"
          }, {
            "name": "param",
            "text": "order - order to apply"
          }]
      }
    },
    "getColumns": {
      "complexType": {
        "signature": "() => Promise<RevoGrid.ColumnRegular[]>",
        "parameters": [],
        "references": {
          "Promise": {
            "location": "global"
          },
          "RevoGrid": {
            "location": "import",
            "path": "../../interfaces"
          }
        },
        "return": "Promise<ColumnRegular[]>"
      },
      "docs": {
        "text": "Receive all columns in data source",
        "tags": []
      }
    },
    "clearFocus": {
      "complexType": {
        "signature": "() => Promise<void>",
        "parameters": [],
        "references": {
          "Promise": {
            "location": "global"
          }
        },
        "return": "Promise<void>"
      },
      "docs": {
        "text": "Clear current grid focus",
        "tags": []
      }
    },
    "getPlugins": {
      "complexType": {
        "signature": "() => Promise<RevoPlugin.Plugin[]>",
        "parameters": [],
        "references": {
          "Promise": {
            "location": "global"
          },
          "RevoPlugin": {
            "location": "import",
            "path": "../../interfaces"
          }
        },
        "return": "Promise<Plugin[]>"
      },
      "docs": {
        "text": "Get all active plugins instances",
        "tags": []
      }
    }
  }; }
  static get elementRef() { return "element"; }
  static get watchers() { return [{
      "propName": "columns",
      "methodName": "columnChanged"
    }, {
      "propName": "theme",
      "methodName": "themeChanged"
    }, {
      "propName": "source",
      "methodName": "dataChanged"
    }, {
      "propName": "pinnedBottomSource",
      "methodName": "dataBottomChanged"
    }, {
      "propName": "pinnedTopSource",
      "methodName": "dataTopChanged"
    }, {
      "propName": "rowDefinitions",
      "methodName": "rowDefChanged"
    }, {
      "propName": "trimmedRows",
      "methodName": "trimmedRowsChanged"
    }, {
      "propName": "grouping",
      "methodName": "groupingChanged"
    }]; }
  static get listeners() { return [{
      "name": "click",
      "method": "handleOutsideClick",
      "target": "document",
      "capture": false,
      "passive": false
    }, {
      "name": "internalRowDragStart",
      "method": "onRowDragStarted",
      "target": undefined,
      "capture": false,
      "passive": false
    }, {
      "name": "internalRowDragEnd",
      "method": "onRowDragEnd",
      "target": undefined,
      "capture": false,
      "passive": false
    }, {
      "name": "internalRowDrag",
      "method": "onRowDrag",
      "target": undefined,
      "capture": false,
      "passive": false
    }, {
      "name": "internalRowMouseMove",
      "method": "onRowMouseMove",
      "target": undefined,
      "capture": false,
      "passive": false
    }, {
      "name": "internalCellEdit",
      "method": "onBeforeEdit",
      "target": undefined,
      "capture": false,
      "passive": false
    }, {
      "name": "internalRangeDataApply",
      "method": "onBeforeRangeEdit",
      "target": undefined,
      "capture": false,
      "passive": false
    }, {
      "name": "internalSelectionChanged",
      "method": "onRangeChanged",
      "target": undefined,
      "capture": false,
      "passive": false
    }, {
      "name": "initialRowDropped",
      "method": "onRowDropped",
      "target": undefined,
      "capture": false,
      "passive": false
    }, {
      "name": "initialHeaderClick",
      "method": "onHeaderClick",
      "target": undefined,
      "capture": false,
      "passive": false
    }, {
      "name": "internalFocusCell",
      "method": "onCellFocus",
      "target": undefined,
      "capture": false,
      "passive": false
    }]; }
}
applyMixins(RevoGridComponent, [GridRenderService]);
